nomad/state/state_store_test.go

```diff
@@ -2901,6 +2901,111 @@ func TestStateStore_DeleteJobTxn_BatchDeletes(t *testing.T) {
 	require.Equal(t, deletionIndex, index)
 }
 
+// TestStatestore_JobTaggedVersion tests that job versions which are tagged
+// do not count against the configured server.job_tracked_versions count,
+// do not get deleted when new versions are created,
+// and *do* get deleted immediately when its tag is removed.
+func TestStatestore_JobTaggedVersion(t *testing.T) {
+	ci.Parallel(t)
+
+	state := testStateStore(t)
+
+	job := mock.MinJob()
+	job.Stable = true
+
+	// add job and fill the versions bucket
+	state.config.JobTrackedVersions = 5
+	for range state.config.JobTrackedVersions {
+		must.NoError(t, state.UpsertJob(structs.MsgTypeTestSetup, nextIndex(state), nil, job.Copy()))
+	}
+
+	// tag versions 0-2
+	for x := range 3 {
+		v := uint64(x)
+		name := fmt.Sprintf("v%d", x)
+		desc := fmt.Sprintf("version %d", x)
+
+		// apply tag
+		must.NoError(t, state.UpdateJobVersionTag(nextIndex(state), job.Namespace, &structs.JobApplyTagRequest{
+			JobID: job.ID,
+			Name:  name,
+			Tag: &structs.JobTaggedVersion{
+				Name:        name,
+				Description: desc,
+			},
+			Version: v,
+		}))
+
+		// confirm
+		got, err := state.JobVersionByTagName(nil, job.Namespace, job.ID, name)
+		must.NoError(t, err)
+		must.Eq(t, name, got.TaggedVersion.Name)
+		must.Eq(t, desc, got.TaggedVersion.Description)
+	}
+
+	// take a little detour to test error conditions with the setup so far
+	t.Run("errors", func(t *testing.T) {
+		// job does not exist
+		err := state.UpdateJobVersionTag(nextIndex(state), job.Namespace, &structs.JobApplyTagRequest{
+			JobID:   "non-existent-job",
+			Tag:     &structs.JobTaggedVersion{Name: "tag name"},
+			Version: 0,
+		})
+		must.ErrorContains(t, err, `job "non-existent-job" version 0 not found`)
+
+		// version does not exist
+		err = state.UpdateJobVersionTag(nextIndex(state), job.Namespace, &structs.JobApplyTagRequest{
+			JobID:   job.ID,
+			Tag:     &structs.JobTaggedVersion{Name: "tag name"},
+			Version: 999,
+		})
+		must.ErrorContains(t, err, fmt.Sprintf("job %q version 999 not found", job.ID))
+
+		// tag name already exists
+		err = state.UpdateJobVersionTag(nextIndex(state), job.Namespace, &structs.JobApplyTagRequest{
+			JobID:   job.ID,
+			Tag:     &structs.JobTaggedVersion{Name: "v0"},
+			Version: 3,
+		})
+		must.ErrorContains(t, err, fmt.Sprintf(`"v0" already exists on a different version of job %q`, job.ID))
+	})
```

Comment:
Yeah it's a bit tricky to hold it all in mind, but I don't believe any of the parent branches/PRs have any test coverage at this layer (I don't see `state_store_test.go` in other PRs), and when I started testing the "don't let tagged versions go away" logic, it felt natural to elaborate here since I'd already done the requisite setup.